### ISSUE TITLE
no-invalid-this: add switch for function name based heuristic 

### ISSUE THREAD

**Issue by andidevi** [eslint/eslint#12271]
* **ESLint Version:** 6.4.0
* **Node Version:** 8.16.1
* **npm Version:** 6.4.1

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**What did you expect to happen?**
Detect invalid this in functions with names with leading uppercase

**What actually happened? Please include the actual, raw output from ESLint.**
The rule no-invalid-this counts functions with leading uppercase in name as constructor.
There is no option to turn this heuristic of.

**Reason**
When writing React Components as a pure function one likes to keep the leading uppercase in the name. This will lead no-invalid-this to ignore uses of this in the pure function, which is a common error when refactoring call components to pure functions.

This is a request to add a switch to turn off the (documented) heuristic decision, that function is a constructor if 
 - the name of the function starts with uppercase, or
 - the function is assigned to a variable which starts with an uppercase letter.

**Comment by mdjermanovic** [comment]
Seems reasonable to me, a project can have a different convention for capitalized functions and not have 'ES5' constructor functions at all.

Also, [User-Defined Components Must Be Capitalized](https://reactjs.org/docs/jsx-in-depth.html#user-defined-components-must-be-capitalized)

**Comment by mdjermanovic** [comment]
I'm willing to champion this, but it still needs one more :thumbsup: from the team.

The name of the option could be `capIsConstructor` (default `true`)?

**Comment by mdjermanovic** [comment]
I removed my vote to avoid confusion, as the issue needs 3 :thumbsup: from other members to be accepted.

To illustrate the idea, the following would be an error in strict mode (e.g., in a module):

```js
/* eslint no-invalid-this: ["error", { "capIsConstructor": false }] */

function Hello() {
  const { foo } = this.state; // Error: Unexpected 'this' 
}
```

Setting the option to `false` says that the project does not use the uppercase convention for constructors and that they would like this rule to treat functions that start with an uppercase as 'regular' functions, i.e. warn if the function that uses `this` isn't assigned to a property etc. - same logic as for all other functions.

**Comment by mysticatea** [comment]
> I removed my vote to avoid confusion, as the issue needs 3 👍 from other members to be accepted.

It's okay even if you don't remove your vote.

Because 👍 on GitHub issues includes votes from the community, it's hard to count votes from the team on issues. Instead, we can check champion and supporters from the team in https://github.com/mysticatea/eslint-evaluating-issues#readme.

**Comment by mdjermanovic** [comment]
I'm working on this.
